File: FundamentalExample/Objective-C/VponFundamentalSampleObjC/VponFundamentalSampleObjC/Sdk/VponSdkNativeTableViewController.m
//
//  VponSdkNativeTableViewController.m
//  VponSdkSampleObjC
//
//  Created by EricChien on 2017/10/3.
//  Copyright © 2017年 Soul. All rights reserved.
//

#import "VponSdkNativeTableViewController.h"
#import "VponSdkNativeCustomTabCell.h"
#import <AdSupport/AdSupport.h>

#import "VpadnNativeAd.h"
#import "VpadnNativeAdsManager.h"
#import "VpadnNativeAdTableViewAdProvider.h"

static NSInteger const kRowStrideForAdCell = 5;
static NSString *const kDefaultCellIdentifier = @"normalIdentifier";
static NSString *const kAdCellIdentifier = @"adIdentifier";

@interface VponSdkNativeTableViewController () <UITableViewDelegate, UITableViewDataSource, VpadnNativeAdsManagerDelegate, VpadnNativeAdDelegate>

@property (strong, nonatomic) VpadnNativeAdsManager *adsManager;
@property (strong, nonatomic) VpadnNativeAdTableViewAdProvider *ads;

@property (strong, nonatomic) NSMutableArray *tableViewContents;

@end

@implementation VponSdkNativeTableViewController

#pragma mark - Lazy Loading
- (NSMutableArray *)tableViewContents {
    if (!_tableViewContents) {
        _tableViewContents = [NSMutableArray array];
        for (NSUInteger i = 0; i < 1000; i++) {
            [_tableViewContents addObject:[NSString stringWithFormat:@"TableView Cell #%d", (int)(i+1)]];
        }
    }
    return _tableViewContents;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 140;
    
    if (!self.adsManager) {
#warning set ad banner id
        self.adsManager = [[VpadnNativeAdsManager alloc] initWithBannerID:@"" forNumAdsRequested:kRowStrideForAdCell];
        self.adsManager.delegate = self;
        [self.adsManager dontUseWVForNA];
        [self.adsManager showTestLog:NO];
    }
    
    [self.adsManager loadAdsWithTestIdentifiers:@[]];
}


#pragma mark - VpadnNativeAdsManagerDelegate

- (void)onVpadnNativeAdsReceived {
    NSLog(@"Ads did loaded");
    VpadnNativeAdsManager *manager = self.adsManager;
    self.adsManager.delegate = nil;
    self.ads = [[VpadnNativeAdTableViewAdProvider alloc] initWithManager:manager];
    self.ads.delegate = self;
    
    [self.tableView reloadData];
}

- (void)onVpadnNativeAdsFailedToLoadWithError:(NSError *)error {
    NSLog(@"Ads did fail with error %@", error);
}

#pragma mark - VpadnNativeAdDelegate
- (void)onVpadnNativeAdPresent:(VpadnNativeAd *)nativeAd {
    NSLog(@"Native Present %@", nativeAd);
}
- (void)onVpadnNativeAdLeaveApplication:(VpadnNativeAd *)nativeAd {
    NSLog(@"Native Leave Application %@", nativeAd);
}
- (void)onVpadnNativeAdDismiss:(VpadnNativeAd *)nativeAd {
    NSLog(@"Native Dismiss %@", nativeAd);
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.ads adjustCount:self.tableViewContents.count forStride:kRowStrideForAdCell] ?:self.tableViewContents.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell;
    if ([self.ads isAdCellAtIndexPath:indexPath forStride:kRowStrideForAdCell]) {
        cell = [tableView dequeueReusableCellWithIdentifier:kAdCellIdentifier forIndexPath:indexPath];
        VpadnNativeAd *ad = [self.ads tableView:tableView nativeAdForRowAtIndexPath:indexPath];
        [(VponSdkNativeCustomTabCell *)cell setNativeAd:ad];
        [ad registerViewForInteraction:cell.contentView withViewController:self];
    } else {
        
        cell = [tableView dequeueReusableCellWithIdentifier:kDefaultCellIdentifier forIndexPath:indexPath];
        NSIndexPath *index = [self.ads adjustNonAdCellIndexPath:indexPath forStride:kRowStrideForAdCell]?: indexPath;
        cell.textLabel.text = [self.tableViewContents objectAtIndex:index.row];
        [cell setBackgroundColor:[UIColor whiteColor]];
    }
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: FrameworkExample/Swift/VponFrameworkSample_swift/VponSdkSampleSwift/Sdk/VponSdkNativeTableViewController.swift
//
//  VponSdkNativeTableViewController.swift
//  VponSdkSampleSwift
//
//  Created by EricChien on 2017/10/3.
//  Copyright © 2017年 EricChien. All rights reserved.
//

import UIKit

import VpadnSDKAdKit

let kRowStrideForAdCell = 5;
let kDefaultCellIdentifier = "normalIdentifier";
let kAdCellIdentifier = "adIdentifier";

extension VponSdkNativeTableViewController: VpadnNativeAdsManagerDelegate {
    func onVpadnNativeAdsReceived() {
        tableView.reloadData()
    }
    
    func onVpadnNativeAdsFailedToLoadWithError(_ error: Error) {
        print("Ads did fail with error \(error.localizedDescription)")
    }
}

extension VponSdkNativeTableViewController: VpadnNativeAdDelegate {
    func onVpadnNativeGet(_ nativeAd: VpadnNativeAd) {
        
    }
    
    func onVpadnNativeAdPresent(_ nativeAd: VpadnNativeAd) {
        print("Native Present \(nativeAd)")
    }
    
    func onVpadnNativeAdLeaveApplication(_ nativeAd: VpadnNativeAd) {
        print("Native Leave Application \(nativeAd)")
    }
    
    func onVpadnNativeAdDismiss(_ nativeAd: VpadnNativeAd) {
        print("Native Dismiss \(nativeAd)")
    }
}

class VponSdkNativeTableViewController: UITableViewController {

    var adsManager : VpadnNativeAdsManager!
    var adsProvider : VpadnNativeAdTableViewAdProvider!
    var tableViewContents = [] as Array
    
    func addContents() {
        tableViewContents = []
        for index in 1...1000 {
            tableViewContents.append("TableView Cell \(index)")
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        addContents()

        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.estimatedRowHeight = 140
        
        if adsManager == nil {
            // TODO: set ad native id
            adsManager = VpadnNativeAdsManager.init(bannerID: "", forNumAdsRequested: UInt(kRowStrideForAdCell))
            adsManager.delegate = self
            adsProvider = VpadnNativeAdTableViewAdProvider.init(manager: self.adsManager)
            adsProvider.delegate = self
        }
        adsManager.loadAds(withTestIdentifiers: [])
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tableViewContents.count + tableViewContents.count / kRowStrideForAdCell
    }

    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if adsManager.isReady() && indexPath.row > 0 && indexPath.row % kRowStrideForAdCell == 0 {
            let cell = tableView.dequeueReusableCell(withIdentifier: kAdCellIdentifier, for: indexPath) as! VponSdkNativeCustomTabCell
            let nativeAd = adsProvider.tableView(tableView, nativeAdForRowAt: indexPath)
            cell.setNativeAd(nativeAd: nativeAd)
            nativeAd.registerView(forInteraction: cell.contentView, with: self)
            return cell
        }
        
        let cell = tableView.dequeueReusableCell(withIdentifier: kDefaultCellIdentifier, for: indexPath)
        let index = indexPath.row - indexPath.row / kRowStrideForAdCell
        cell.textLabel?.text = tableViewContents[index] as? String
        cell.backgroundColor = UIColor.white
        return cell
    }


}
